cub/device/device_batch_memcpy.cuh

```diff
@@ -0,0 +1,94 @@
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+/**
+ * \file
+ * cub::DeviceBatchMemcpy provides device-wide, parallel operations for copying data from a number of given source
+ * buffers to their corresponding destination buffer.
+ */
+
+#pragma once
+
+#include "../config.cuh"
+#include "dispatch/dispatch_batch_memcpy.cuh"
+
+#include <cstdint>
+#include <type_traits>
+
+CUB_NAMESPACE_BEGIN
+
+/**
+ * @brief Copies data from a batch of given source buffers to their corresponding destination buffer.
+ * @note If any input buffer aliases memory from any output buffer the behavior is undefined. If any output buffer
+ * aliases memory of another output buffer the behavior is undefined. Input buffers can alias one another.
+ *
+ * @tparam InputBufferIt <b>[inferred]</b> Device-accessible random-access input iterator type providing the pointers to
+ * the source memory buffers
+ * @tparam OutputBufferIt <b>[inferred]</b> Device-accessible random-access input iterator type providing the pointers
+ * to the destination memory buffers
+ * @tparam BufferSizeIteratorT <b>[inferred]</b> Device-accessible random-access input iterator type providing the
+ * number of bytes to be copied for each pair of buffers
+ * @param d_temp_storage [in] Device-accessible allocation of temporary storage.  When NULL, the required allocation
+ * size is written to \p temp_storage_bytes and no work is done.
+ * @param temp_storage_bytes [in,out] Reference to size in bytes of \p d_temp_storage allocation
+ * @param input_buffer_it [in] Iterator providing the pointers to the source memory buffers
+ * @param output_buffer_it [in] Iterator providing the pointers to the destination memory buffers
```

Comment:
Is it sufficiently obvious in CUB APIs that iterators should always be device accessible?

Reply:
Good point 👍  Better to be explicit about it. I'll add it to the documentation.